### ISSUE TITLE
Allow BigInt and BigFloat to be initialized from a byte slice

### ISSUE THREAD

**Issue by watzon** [crystal-lang/crystal#10101]
Currently it's possible to initialize a BigInt and BigFloat using a string, but not a byte slice (as far as I know). This means that it would be possible to take the byte slice, convert it to a string, and then initialize the big number, but it's not very efficient to create that intermediate string every time. From what I've seen from the BigInt source code it should be possible to add the following initializer:

```crystal
def initialize(bytes : Bytes, base = 10)
  err = LibGMP.init_set_str(out @mpz, bytes, base)
  if err == -1
    raise ArgumentError.new("Invalid BigInt: #{bytes}")
  end
end
```

I would be happy to make a PR

**Comment by j8r** [comment]
I don't think there is much overhead, because the slice buffer is already allocated. There is just the class overhead of String.

If a Slice is taken, it can be confused with a ByteFormat.

**Comment by straight-shoota** [comment]
@j8r `String` always copies the buffer to ensure immutability, so there's definitely an unncessary overhead.
Adding a `Bytes` overload looks fine.

**Comment by j8r** [comment]
@straight-shoota I thought it was copy-on-write; why copying if there is no modification?

You're right, a copy is done https://github.com/crystal-lang/crystal/blob/12bf665/src/string.cr#L210.

**Comment by j8r** [comment]
I got my answer, copying prevents the string to be modified if the Bytes buffer do.

**Comment by asterite** [comment]
I think a Bytes overload is fine. It's a bit too low level, but it doesn't hurt having that either.

**Comment by watzon** [comment]
@j8r I was actually confused on what I was looking to do when I originally created this issue, although it does make sense to me to have a way to make a `BigInt` directly from a `Slice(UInt8)` (where that slice is just a string in byte form).

It would also be nice though (and this probably deserves its own issue) to be able to use `IO::ByteFormat` to encode/decode BigInts. I've already come up with a proof of concept that uses left shifts and bitwise ORs to create a BigInt from a list of bytes. It would be a nice thing to have in the standard library if someone can make it more concise/efficient.

**Comment by asterite** [comment]
there's mpz_export and mpz_import for that.

**Comment by watzon** [comment]
@asterite any particular why that hasn't already been done then? Outside of time/priority constraints? If not I could probably try to PR something.

**Comment by asterite** [comment]
I don't use BigInt and bigint is never a part of any protocol.

**Comment by watzon** [comment]
I'll PR it then. I need it for a project I'm working on.

**Comment by asterite** [comment]
Please don't. I don't think I'll approve it. This is very specific to libgmp.If you need it you can do it as a shard or for your project.

**Comment by j8r** [comment]
Interestingly in JS;
```js
let buffer = new ArrayBuffer(16);
let dataview = new DataView(buffer);

// Highest possible BigInt value that fits in a signed 64-bit integer
const max = 2n ** (64n - 1n) - 1n;
dataview.setBigInt64(0, max);

console.log(Array.from(new Uint8Array(buffer, 0, BigInt64Array.BYTES_PER_ELEMENT)))
```
Result: `[127, 255, 255, 255, 255, 255, 255, 255]`

I wonder what format it is, and if it is standard.

**Comment by straight-shoota** [comment]
Some research on big number interchange format would probably help. Maybe GMP byte format is somewhat standard. Or something different is. Maybe it's just string, though.

**Comment by j8r** [comment]
So, for JS, per https://tc39.es/proposal-bigint/, a 64-bit two's complement signed integer is used. This is similar as encoding in Crystal a `Int64` in bytes with `IO::ByteFormat::BigEndian`:
```cr
io = IO::Memory.new
io.write_bytes(9223372036854775807_i64, IO::ByteFormat::BigEndian)
p io.to_slice #=> Bytes[127, 255, 255, 255, 255, 255, 255, 255]
```
Note that is JS, 64bit integers are not natively supported, only up to 53 bits (https://en.wikipedia.org/wiki/IEEE_754-2008_revision).

[There are different ways of representing arbitrary-precision numbers](https://en.wikipedia.org/wiki/Arbitrary-precision_arithmetic#Software_libraries), which vary depending of libraries.

A shard is better fitted, I propose to close this issue. This cannot be in the stdlib, because multiple representations are possible. Furthermore, as @asterite said, this use-case is not frequent enough.

**Comment by oprypin** [comment]
Um well the issue has just been derailed from the original proposal, first it wasn't about a serialization format, but about doing literally the same thing as `initialize(String)`, just without requiring it to be a `String` (pass `Bytes` instead). Which nobody objects to.

I do have a small objection, though: people might confuse this kind of initialization with actual deserialization from a byte format, as we've done in this very issue. And then further confusion would ensue if we ever introduced an actual serialization format.

**Comment by j8r** [comment]
Yes, that what [I was afraid initially](https://github.com/crystal-lang/crystal/issues/10101#issuecomment-748469748).

I suggest using something like `.from_utf8(bytes : Bytes)`, like done for [String.from_utf16(slice : Slice(UInt16))](https://crystal-lang.org/api/master/String.html#from_utf16(slice:Slice(UInt16)):String-class-method) to avoid any confusion.

**Comment by straight-shoota** [comment]
Related to  #9299
